First pass on porting /tg/ maploader functionality

thasc opened the pull request:
🆑 thasc
rscadd: Admins can now spawn map templates (and upload .dmm files to spawn) via the Fun tab.
/🆑

This pass of porting the map loader is *just* about:

- Allowing people to contribute map templates in the form of a .dmm file and a brief datum (see `code/datums/map_templates`).
  - (For example, you (yes, *you!*) might create a nice little house, put its .dmm in with the other exoplanet map templates, and write a map datum referencing it.)
- Allowing admins to place down templates, either in the existing zlevels, or as a new one (or group of zlevels if multi-level).
  - (That means that nice little house you created might turn up in the ship's hangar as an admin fat-fingered the button!)
- Exposing the `seedRuins` function, which randomly spawns templates (either exoplanet or space).
  - (That means the next exoplanet you visit might be filled up with *many* of your house!)
    - (Although not at the moment, because it isn't hooked in anywhere.)

So there is currently no work done in this PR for:

- Changing the main map by vote. (This will require lots of work.)
- Conditionally spawning a subset of all away missions.

Next pass will be about:

- Automatically picking and spawning away sites at runtime so we don't need them compiled in.

Notes:

- `seedRuins` currently isn't called from anywhere.
- Away sites aren't spawned by this system yet, either.
- Therefore there is currently no actual purpose in dividing things into exoplanet ruins and space ruins and away sites.
  - But there *will* be, so do it!
- Object initialisation can be sketchy, but usually that can be resolved. If you've made a .dmm with some object on it, and that object doesn't seem to be initialising right when spawned by template, check to see if its superclass constructor call in `New` is happening at the very start, or at all. If it's not, it might not initialise right.

I haven't done something this big before, and it's a yank from /tg/, so please comment the heck out of it. Note that while `reader.dm` is a lot of changes, it's basically a direct port from /tg/, so it's theoretically pretty proven.

thasc committed:
First pass on porting /tg/ maploader.

thasc commented on code/__defines/misc.dm:
```diff
@@ -1,8 +1,10 @@
 #define DEBUG
 // Turf-only flags.
 #define NOJAUNT 1 // This is used in literally one place, turf.dm, to block ethereal jaunt.
+#define NORUINS 2
```
We don't use this anywhere, but if applied as a turf flag, it will prevent ruins from being made on it.

thasc commented on code/_helpers/global_access.dm:
```diff
@@ -51,16 +51,18 @@
 		if("SSprocessing")
 			return global.SSprocessing;
+		if("SSvines")
+			return global.SSvines;
```
Yeah, I dunno either. Tools likes to nudge stuff around, I guess!

thasc commented on code/controllers/master_controller.dm:
```diff
@@ -65,32 +56,10 @@ datum/controller/game_controller/proc/setup_objects()
-	report_progress("Initializing pipe networks")
-	for(var/obj/machinery/atmospherics/machine in SSmachines.machinery)
-		machine.build_network()
-		CHECK_SLEEP_MASTER
```
All this red has been nudged into the machines subsystem instead, and set up so that we can initialise an explicit list of machinery, rather than everything (useful for templates).

thasc commented on code/controllers/subsystems/atoms.dm:
```diff
@@ -21,7 +21,7 @@ SUBSYSTEM_DEF(atoms)
 	return ..()
 
-/datum/controller/subsystem/atoms/proc/InitializeAtoms(list/atoms)
+/datum/controller/subsystem/atoms/proc/InitializeAtoms(list/atoms, do_report_progress=TRUE)
```
Yuck, but.

PsiOmegaDelta replied:
Why would you not want to report progress even if late-loading?

thasc replied:
I think I interpreted it as ultimately writing out onto `world` and bugging everyone. Looks like that's not actually the case, so I'll yank all this.

thasc commented on code/controllers/subsystems/machines.dm:
```diff
@@ -80,13 +81,54 @@ if(current_step == this_step || (check_resumed && !resumed)) {\
+#endif
+
+#undef CHECK_SLEEP_MACHINES
+
```
And this is where that stuff from the master controller ended up.

thasc commented on code/controllers/subsystems/mapping.dm:
```diff
@@ -0,0 +1,49 @@
+SUBSYSTEM_DEF(mapping)
```
So this guy has had a whole load of shit taken out of it. The real /tg/ map loader supports switching maps, but that's currently a heck of an endeavour to engage in, so I've excised all that functionality in favour of the bare essentials for the moment.

thasc commented on code/datums/map_templates/exoplanet_ruins.dm:
```diff
@@ -0,0 +1,14 @@
+	id = "exoplanet_little_house"
+	description = "oh wow look it's a cute little house"
+	suffix = "little_house.dmm"
+	cost = 1
```
**This** is an example of a map template. Every new template added should have one of these guys. They'll be automatically picked up by the map loader and stored as options for admins to spawn via `Debug/Map Template - Place` or `Debug/Map Template - Place In New Z`.

thasc commented on code/game/atoms.dm:
```diff
@@ -25,6 +25,9 @@
 
+	if(GLOB.use_preloader && (src.type == GLOB._preloader.target_path))//in case the instanciated atom is creating other atoms in New()
+		GLOB._preloader.load(src)
+
```
This guy is (only hopefully) the first actual bit of real work done in any object's `New`. It helps ensure an object's properties (custom name, etc) are picked up out of the template dmm and written into the object *before* anything starts working with them. That only happens, though, if the superclass constructor is the first thing in every `New`. Which was not the case for APCs, as you'll see by the change this PR made to that type.

thasc commented on code/game/objects/effects/landmarks.dm:
```diff
@@ -241,4 +241,18 @@
+	name = "ruin_[GLOB.ruin_landmarks.len + 1]"
+	..(loc)
+	ruin_template = my_ruin_template
+	GLOB.ruin_landmarks |= src
```
Mostly used by `seedRuins`.

thasc commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+			log_and_message_admins("has placed a map template ([template.name]) at [T.loc].")
+		else
+			to_chat(usr, "Failed to place map")
+	usr.client.images -= preview
```
Behaviour of `Place` - allows selection of a template, shows green overlay to indicate where it will spawn, and then offers confirmation before loading. Supports multi-z templates. If the template is too tall for where you are (e.g. you're spawning a three-z-tall template on the Bridge deck), it'll just elide any zlevels that can't fit.

thasc commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+	if (new_z_centre)
+		log_and_message_admins("has placed a map template ([template.name]) on a new zlevel.", location=new_z_centre)
+	else
+		to_chat(usr, "Failed to place map")
```
Behaviour of `Place In New Z` - similar to before. No green overlay, no confirmation. (Might add confirmation.) Spawns the template in a new zlevel. If it's a multi-z template, it will refuse to spawn it if there's not enough 'space' left in the zlevel system (e.g. 17 for `basic.dm`).

thasc commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+		SSmapping.map_templates[M.name] = M
+		message_admins("<span class='adminnotice'>[key_name_admin(usr)] has uploaded a map template ([map])</span>")
+	else
+		to_chat(usr, "Map template '[map]' failed to load properly")
```
Behaviour of `Upload` - filepicker. Pick a .dmm file. Now it's a template. Ez pz!

thasc commented on code/modules/maps/map_template.dm:
```diff
@@ -0,0 +1,103 @@
+
+	SSatoms.InitializeAtoms(atoms, FALSE)
+	SSmachines.setup_powernets_for_cables(cables)
+	SSmachines.setup_atmos_machinery(atmos_machines, FALSE)
```
Above is quite an important function. This one initialises all the stuff placed in your template by the map loader - every atom, every machine, powernet, atmos, the lot.

thasc commented on code/modules/maps/map_template.dm:
```diff
@@ -0,0 +1,103 @@
+	initTemplateBounds(bounds)
+	log_game("Z-level [name] loaded at [x],[y],[world.maxz]")
+
+	return locate(world.maxx/2, world.maxy/2, world.maxz)
```
Spawn a template in the middle of a new zlevel(s).

thasc commented on code/modules/maps/reader.dm:
```diff
@@ -3,8 +3,21 @@
 //////////////////////////////////////////////////////////////
```
This file, responsible for reading in and parsing .dmms, is mostly a direct port from /tg/, so it's pretty battle-proven. I'll comment the exception where I wrote some stuff.

thasc commented on code/modules/maps/reader.dm:
```diff
@@ -16,86 +29,167 @@ var/global/dmm_suite/preloader/_preloader = null
+			var/is_on_an_existing_zlevel = zcrd <= world.maxz
+
+			if (is_on_an_existing_zlevel && !is_connected_to_lower_levels)
+				continue
```
96 to 100 are my addition, added to slice off the tops of templates if they're too tall to fit in a collection of linked zlevels. Otherwise we'd get the arses of templates in Torch and their heads in Centcomm.

thasc commented on code/modules/maps/ruins.dm:
```diff
@@ -0,0 +1,104 @@
+/proc/seedRuins(list/z_levels = null, budget = 0, whitelist = /area/space, list/potentialRuins)
```
This guy is cool. Give it a budget and a list of possible 'ruin' (read: tiny map) templates, point it at some zlevels, and it'll randomly throw down random ruins from that list until it runs out of space or budget. When exoplanets are behaving themselves again it'd be great to integrate this into their initialisations.

thasc commented on code/modules/multiz/basic.dm:
```diff
@@ -1,9 +1,12 @@
+GLOBAL_VAR_CONST(MAX_CONNECTABLE_ZLEVEL_INDEX, 17)
```
I wanted to expose the number of zlevels 'available' in the system in the least dumb looking way. Tell me if I didn't succeed.

PsiOmegaDelta replied:
Technically this should be named something akin to "MAX_NUMBER_OF_ZLEVELS", as it is the maximum number of Z-levels in total, not maximum number of connections between them.

thasc replied:
My understanding of this system is that, while we can push out further than 17 zlevels, any zlevels beyond index 17 cannot be involved in zlevel connections. So it's not a *maximum number of zlevels*, exactly, but the highest index that can be connected. That's why I picked that name. Does that make sense?

PsiOmegaDelta replied:
So it's a HIGHEST_CONNECTABLE_ZLEVEL_INDEX?

thasc replied:
That's a good name too!

thasc commented on code/modules/power/apc.dm:
```diff
@@ -193,8 +195,6 @@
 		src.update_icon()
 
-	..()
-
```
Moved to the start of `New` to initialise things properly when loaded as a template object.

comma reviewed (dismissed):
Nothing really jumped out at me after reading through.

thasc commented:
Updooted with fresh sexy globals hash.

I should note that there are three new config files (*_blacklist.txt) that will need to be present in the config folder.

thasc commented:
CI's really not my friend today. I'll look tomorrow.

PsiOmegaDelta commented on code/controllers/subsystems/machines.dm:
```diff
@@ -80,13 +81,54 @@ if(current_step == this_step || (check_resumed && !resumed)) {\
+#define CHECK_SLEEP_MACHINES if(!(initialization_stage & INITIALIZATION_NOW) && ++initialized_objects > 500) { initialized_objects=0;sleep(world.tick_lag); }
+#endif
+
+datum/controller/subsystem/machines/proc/setup_atmos_machinery(list/machines, do_report_progress=TRUE)
```
Multiple init calls on atmospherics objects has been known to more or less subtly break things before. Have you confirmed that all atmospherics objects remain functional?

thasc replied:
I'll reconfirm with a proper look. The calls to this - one during round initialisation, and others during initialisation of templates - apply to different collections of objects. I'll try to figure out cases where that would fail.

PsiOmegaDelta commented on code/controllers/subsystems/machines.dm:
```diff
@@ -80,13 +81,54 @@ if(current_step == this_step || (check_resumed && !resumed)) {\
+#define CHECK_SLEEP_MACHINES if(!(initialization_stage & INITIALIZATION_NOW) && ++initialized_objects > 500) { initialized_objects=0;sleep(world.tick_lag); }
+#endif
+
+datum/controller/subsystem/machines/proc/setup_atmos_machinery(list/machines, do_report_progress=TRUE)
```
And again, is conditional progress reporting required?

thasc replied:
Nope, as above, looks like I did a dumb.

PsiOmegaDelta commented on code/controllers/subsystems/machines.dm:
```diff
@@ -80,13 +81,54 @@ if(current_step == this_step || (check_resumed && !resumed)) {\
+		report_progress("Initializing atmos machinery")
+	for(var/obj/machinery/atmospherics/A in machines)
+		A.atmos_init()
+		CHECK_SLEEP_MACHINES
```
Please don't use the hacky sleep check in subsystems.
Make `CHECK_TICK`, similar as in the atoms subsystem.

thasc replied:
That *does* look better... didn't spot that before. I'll sort that. Thanks!

PsiOmegaDelta commented on code/game/objects/effects/landmarks.dm:
```diff
@@ -241,4 +241,18 @@
+	var/datum/map_template/ruin/ruin_template
+
+/obj/effect/landmark/ruin/New(loc, my_ruin_template)
+	name = "ruin_[GLOB.ruin_landmarks.len + 1]"
```
This will potentially allow multiple landmarks with the same name, if ruin landmarks can ever be added after another has been deleted.
`"ruin_[sequential_id(/obj/effect/landmark/ruin)]"` would be a safer naming scheme.

thasc replied:
Sounds good.

PsiOmegaDelta commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+/client/proc/map_template_load()
+	set category = "Debug"
+	set name = "Map Template - Place"
```
Lacking admins permission checks.
If only you had used the datum based admin action system instead.

thasc replied:
I wasn't aware that existed. Perils of porting stuff from other codebases, I guess. Need to know what bits need to change to fit the local flavour. I'll sort it.

PsiOmegaDelta commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+	usr.client.images += preview
+	if(alert(usr,"Confirm location.","Template Confirm","Yes","No") == "Yes")
+		if(template.load(T, centered = TRUE))
+			log_and_message_admins("has placed a map template ([template.name]) at [T.loc].")
```
log_and_message_admins() already adds the current location

PsiOmegaDelta commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+
+/client/proc/map_template_load_new_z()
+	set category = "Debug"
+	set name = "Map Template - Place In New Z"
```
More missing permission checks.

PsiOmegaDelta commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,58 @@
+
+/client/proc/map_template_upload()
+	set category = "Debug"
+	set name = "Map Template - Upload"
```
Missing permission checks.

PsiOmegaDelta commented on code/modules/multiz/disabled.dm:
```diff
@@ -1,3 +1,4 @@
+GLOBAL_VAR_CONST(MAX_CONNECTABLE_ZLEVEL_INDEX, 0)
```
Without Multi-Z you can technically have as many Z-levels as BYOND can support, you just don't get full interaction between Z-levels.

thasc commented:
Done some updates - two review comments not covered are around checking atmospheric initialisation is fine (no issue seen so far but I'll confirm and try to prove over the next day or so), and I haven't renamed that global const defining the largest z-index that can be used by our multi-z systems.

thasc commented:
So for atmospherics, all of the Torch stuff tests fine and works as expected.

There is a case I can imagine where an atmos machine could be initialised twice - when a map template is written into an location that already contains atmos machines.

That is, if you blap a map template down in the middle of atmospherics, the initialisation of that template will first identify its cuboid outer bounds, and then call `atmos_init` etc on all of the filters and mixers and pumps and yadda yadda within that cuboid.

Simplest option might be to nuke all existing `/obj` and `/mob` within those bounds as a pre-step. Another simple option is to deny template placement if the bounds would contain sensitive equipment. Another is to ensure reinitialisation is harmless.

Lohikar commented:
Just a heads up: the unpatched /tg/ maploader cannot handle simple lists of non-strings, like req_access (unless this has been changed to not be a list).

thasc commented:
Thanks for the heads up, I see /tg/ dmms use `req_access_txt` instead?

Loading our main map with this maploader is a long way off right now, so it's not a showstopper, but I'll do some testing with the templates, and if they misbehave with lists I'll have to police any new templates. Or set up a unit test of some kind.

Lohikar commented:
Alternately, Aurora has a modified version of this code that can handle these lists.

Shouldn't be a big issue if you're just using this for simpler templates though.

thasc commented:
Did a quick test (man is it ever nice to be able to edit dmms and reload them without having to take the server down to recompile) was able to reproduce the issue you mentioned. Thanks again for the heads-up, I'll take a gander at Aurora's implementation. 👌

Lohikar commented:
I should note that the list handling is in Aurora's `development` branch, not `master`.

PsiOmegaDelta commented on code/modules/admin/verbs/map_template_loadverb.dm:
```diff
@@ -0,0 +1,67 @@
+	set desc = "Pick a map template to load at your current location. You will be able to confirm bounds before committing."
+	set name = "Map Template - Place"
+
+	if (!check_rights(R_DEBUG)) return
```
Plain admins are not allowed to load maps?

thasc replied:
If plain admins don't have the R_DEBUG right, then no.

I'm open to alternatives. Debug feels weird for this. Build, maybe? Or fun?

Lohikar commented on code/modules/maps/reader.dm:
```diff
@@ -16,86 +29,167 @@ var/global/dmm_suite/preloader/_preloader = null
+				for(var/t in block(locate(bounds[MAP_MINX], bounds[MAP_MINY], bounds[MAP_MINZ]), locate(bounds[MAP_MAXX], bounds[MAP_MAXY], bounds[MAP_MAXZ])))
+					var/turf/T = t
+					//we do this after we load everything in. if we don't; we'll have weird atmos bugs regarding atmos adjacent turfs
+					GLOB.universe.OnTurfChange(T)
```
Surely this should be `/turf/post_change()`. This proc is completely unrelated afaik.

comma replied:
Yeah, I don't think any universe states actually even implement it.

comma commented on code/modules/maps/reader.dm:
```diff
@@ -106,7 +106,7 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new)
 					world.maxz = zcrd //create a new z_level if needed
 				if(!no_changeturf)
-					WARNING("Z-level expansion occurred without no_changeturf set, this may cause problems when /turf/AfterChange is called")
+					WARNING("Z-level expansion occurred without no_changeturf set, this may cause problems when universe.OnTurfChange is called.")
```
Ditto, universe turf changing is different beastie

thasc commented:
Ok, so initialisation is now occurring just to the atoms the map loader created, rather than everything in the thing's bounding box. That should prevent any weird re-initialisation issues for initialisation-sensitive types.

I've added a commit to create lighting overlays on new zlevels - AFAICT that's the correct way to do it, but if there's a better way let me know.

I've added the option of deleting certain kinds of atom from the template's turfs before laying it down. This is to enable admins to put templates on exoplanets without then having to go through and delete all the flowers (or garbage or whatever) from inside the template.

If everything makes sense I'll squash, until then it's better to have the commits separate.

PsiOmegaDelta commented:
Where are the calls to `SSatoms.map_loader_begin()` and `SSatoms.map_loader_stop()` to prevent map-created objects from being initialized twice?

PsiOmegaDelta reviewed (changes requested):
See question above.

thasc commented:
They're in `reader.dm` - Github likes to hide large diffs by default, so it probably evaded a ctrl-f that way.

thasc commented:
Squashed. Dismissed all reviews automatically as a result. Good work me?
